File: kits/data-integration/providers/transform_plugin/StripTags.swift
// Transform Plugin Provider: strip_tags
// Remove HTML tags with optional allowlist and entity decoding.
// See Architecture doc for transform plugin interface contract.

import Foundation

public final class StripTagsTransformProvider {
    public static let providerId = "strip_tags"
    public static let pluginType = "transform_plugin"

    public init() {}

    public func transform(value: Any, config: TransformConfig) throws -> Any {
        if value is NSNull { return NSNull() }

        var html = String(describing: value)
        let allowedTags = (config.options["allowedTags"] as? [String]) ?? []
        let decodeEntities = (config.options["decodeEntities"] as? Bool) ?? true
        let preserveWhitespace = (config.options["preserveWhitespace"] as? Bool) ?? true

        // Insert whitespace for block elements
        if preserveWhitespace {
            let blockTags = ["p", "div", "br", "hr", "h1", "h2", "h3", "h4", "h5", "h6",
                           "ul", "ol", "li", "table", "tr", "td", "th",
                           "blockquote", "pre", "section", "article"]
            for tag in blockTags {
                if !allowedTags.contains(tag) {
                    if let regex = try? NSRegularExpression(pattern: "<\(tag)[^>]*>", options: .caseInsensitive) {
                        html = regex.stringByReplacingMatches(
                            in: html, range: NSRange(html.startIndex..., in: html), withTemplate: " ")
                    }
                    html = html.replacingOccurrences(of: "</\(tag)>", with: " ",
                        options: .caseInsensitive)
                }
            }
            if !allowedTags.contains("br") {
                html = html.replacingOccurrences(of: "<br>", with: "\n", options: .caseInsensitive)
                html = html.replacingOccurrences(of: "<br/>", with: "\n", options: .caseInsensitive)
                html = html.replacingOccurrences(of: "<br />", with: "\n", options: .caseInsensitive)
            }
        }

        // Remove HTML comments
        if let regex = try? NSRegularExpression(pattern: "<!--[\\s\\S]*?-->") {
            html = regex.stringByReplacingMatches(
                in: html, range: NSRange(html.startIndex..., in: html), withTemplate: "")
        }

        // Strip tags
        if allowedTags.isEmpty {
            if let regex = try? NSRegularExpression(pattern: "<[^>]+>") {
                html = regex.stringByReplacingMatches(
                    in: html, range: NSRange(html.startIndex..., in: html), withTemplate: "")
            }
        } else {
            let allowedSet = Set(allowedTags.map { $0.lowercased() })
            let pattern = "</?([a-zA-Z][a-zA-Z0-9]*)\\b[^>]*/?"
            if let regex = try? NSRegularExpression(pattern: pattern + ">") {
                let nsHtml = html as NSString
                let matches = regex.matches(in: html, range: NSRange(location: 0, length: nsHtml.length))
                var result = html
                for match in matches.reversed() {
                    let tagName = nsHtml.substring(with: match.range(at: 1)).lowercased()
                    if !allowedSet.contains(tagName) {
                        result = (result as NSString).replacingCharacters(in: match.range, with: "")
                    }
                }
                html = result
            }
        }

        // Decode entities
        if decodeEntities {
            html = decodeHtmlEntities(html)
        }

        // Normalize whitespace
        html = html.components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }
            .joined(separator: " ")

        return html.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func decodeHtmlEntities(_ text: String) -> String {
        var result = text
        let entities: [(String, String)] = [
            ("&amp;", "&"), ("&lt;", "<"), ("&gt;", ">"), ("&quot;", "\""),
            ("&#39;", "'"), ("&apos;", "'"), ("&nbsp;", " "),
            ("&mdash;", "\u{2014}"), ("&ndash;", "\u{2013}"),
            ("&hellip;", "\u{2026}"), ("&bull;", "\u{2022}"),
            ("&copy;", "\u{00A9}"), ("&reg;", "\u{00AE}"),
            ("&trade;", "\u{2122}"), ("&times;", "\u{00D7}"),
            ("&euro;", "\u{20AC}"), ("&pound;", "\u{00A3}"),
        ]
        for (entity, char) in entities {
            result = result.replacingOccurrences(of: entity, with: char)
        }

        // Decimal numeric entities
        if let regex = try? NSRegularExpression(pattern: "&#(\\d+);") {
            let nsResult = result as NSString
            let matches = regex.matches(in: result, range: NSRange(location: 0, length: nsResult.length))
            for match in matches.reversed() {
                let numStr = nsResult.substring(with: match.range(at: 1))
                if let code = UInt32(numStr), let scalar = Unicode.Scalar(code) {
                    result = (result as NSString).replacingCharacters(
                        in: match.range, with: String(Character(scalar)))
                }
            }
        }

        return result
    }

    public func inputType() -> TypeSpec {
        return TypeSpec(type: "string", nullable: true)
    }

    public func outputType() -> TypeSpec {
        return TypeSpec(type: "string", nullable: true)
    }
}

File: repertoire/concepts/data-integration/providers/transform_plugin/strip_tags.rs
// Transform Plugin Provider: strip_tags
// Remove HTML tags with optional allowlist and entity decoding.
// See Architecture doc for transform plugin interface contract.

use std::collections::HashMap;
use std::collections::HashSet;

pub const PROVIDER_ID: &str = "strip_tags";
pub const PLUGIN_TYPE: &str = "transform_plugin";

#[derive(Debug, Clone)]
pub enum Value {
    String(String),
    Number(f64),
    Integer(i64),
    Boolean(bool),
    Null,
    Array(Vec<Value>),
    Object(HashMap<String, Value>),
}

#[derive(Debug, Clone)]
pub struct TransformConfig {
    pub options: HashMap<String, Value>,
}

#[derive(Debug, Clone)]
pub struct TypeSpec {
    pub type_name: String,
    pub nullable: bool,
}

#[derive(Debug)]
pub enum TransformError {
    InvalidInput(String),
}

impl std::fmt::Display for TransformError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TransformError::InvalidInput(msg) => write!(f, "Invalid input: {}", msg),
        }
    }
}

pub struct StripTagsTransformProvider;

impl StripTagsTransformProvider {
    pub fn new() -> Self {
        Self
    }

    pub fn transform(&self, value: &Value, config: &TransformConfig) -> Result<Value, TransformError> {
        let html = match value {
            Value::Null => return Ok(Value::Null),
            Value::String(s) => s.clone(),
            _ => return Err(TransformError::InvalidInput("Expected string input".to_string())),
        };

        let allowed_tags: HashSet<String> = match config.options.get("allowedTags") {
            Some(Value::Array(arr)) => {
                arr.iter().filter_map(|v| {
                    if let Value::String(s) = v { Some(s.to_lowercase()) } else { None }
                }).collect()
            }
            _ => HashSet::new(),
        };

        let decode_entities = match config.options.get("decodeEntities") {
            Some(Value::Boolean(false)) => false,
            _ => true,
        };

        let preserve_whitespace = match config.options.get("preserveWhitespace") {
            Some(Value::Boolean(false)) => false,
            _ => true,
        };

        let mut result = html;

        // Insert whitespace for block elements before stripping
        if preserve_whitespace {
            let block_tags = [
                "p", "div", "h1", "h2", "h3", "h4", "h5", "h6",
                "ul", "ol", "li", "table", "tr", "td", "th",
                "blockquote", "pre", "section", "article",
            ];
            for tag in &block_tags {
                if !allowed_tags.contains(*tag) {
                    // Replace opening tags with space
                    let open = format!("<{}", tag);
                    while let Some(pos) = result.find(&open) {
                        if let Some(end) = result[pos..].find('>') {
                            result = format!("{} {}", &result[..pos], &result[pos + end + 1..]);
                        } else {
                            break;
                        }
                    }
                    // Replace closing tags with space
                    let close = format!("</{}>", tag);
                    result = result.replace(&close, " ");
                }
            }
            // <br> to newline
            if !allowed_tags.contains("br") {
                result = result.replace("<br>", "\n")
                    .replace("<br/>", "\n")
                    .replace("<br />", "\n");
            }
        }

        // Remove HTML comments
        while let Some(start) = result.find("<!--") {
            if let Some(end) = result[start..].find("-->") {
                result = format!("{}{}", &result[..start], &result[start + end + 3..]);
            } else {
                break;
            }
        }

        // Strip tags (preserving allowlisted ones)
        if allowed_tags.is_empty() {
            result = self.strip_all_tags(&result);
        } else {
            result = self.strip_tags_except(&result, &allowed_tags);
        }

        // Decode HTML entities
        if decode_entities {
            result = self.decode_html_entities(&result);
        }

        // Normalize whitespace
        result = self.normalize_whitespace(&result);

        Ok(Value::String(result.trim().to_string()))
    }

    fn strip_all_tags(&self, html: &str) -> String {
        let mut result = String::new();
        let mut in_tag = false;
        for ch in html.chars() {
            if ch == '<' {
                in_tag = true;
            } else if ch == '>' {
                in_tag = false;
            } else if !in_tag {
                result.push(ch);
            }
        }
        result
    }

    fn strip_tags_except(&self, html: &str, allowed: &HashSet<String>) -> String {
        let mut result = String::new();
        let mut remaining = html;

        while let Some(pos) = remaining.find('<') {
            result.push_str(&remaining[..pos]);
            remaining = &remaining[pos..];

            if let Some(end) = remaining.find('>') {
                let tag = &remaining[..end + 1];
                let tag_name = self.extract_tag_name(tag);

                if allowed.contains(&tag_name.to_lowercase()) {
                    result.push_str(tag);
                }
                remaining = &remaining[end + 1..];
            } else {
                result.push_str(remaining);
                break;
            }
        }
        result.push_str(remaining);
        result
    }

    fn extract_tag_name(&self, tag: &str) -> String {
        let inner = tag.trim_start_matches('<').trim_start_matches('/');
        let mut name = String::new();
        for ch in inner.chars() {
            if ch.is_alphanumeric() {
                name.push(ch);
            } else {
                break;
            }
        }
        name
    }

    fn decode_html_entities(&self, text: &str) -> String {
        let mut result = text.to_string();
        let entities = [
            ("&amp;", "&"), ("&lt;", "<"), ("&gt;", ">"), ("&quot;", "\""),
            ("&#39;", "'"), ("&apos;", "'"), ("&nbsp;", " "),
            ("&mdash;", "\u{2014}"), ("&ndash;", "\u{2013}"),
            ("&hellip;", "\u{2026}"), ("&bull;", "\u{2022}"),
            ("&copy;", "\u{00A9}"), ("&reg;", "\u{00AE}"),
            ("&trade;", "\u{2122}"), ("&times;", "\u{00D7}"),
            ("&euro;", "\u{20AC}"), ("&pound;", "\u{00A3}"),
        ];
        for (entity, replacement) in &entities {
            result = result.replace(entity, replacement);
        }

        // Decode decimal numeric entities (&#NNN;)
        while let Some(start) = result.find("&#") {
            if let Some(end) = result[start..].find(';') {
                let num_str = &result[start + 2..start + end];
                if let Ok(code) = num_str.parse::<u32>() {
                    if let Some(ch) = char::from_u32(code) {
                        result = format!("{}{}{}", &result[..start], ch, &result[start + end + 1..]);
                        continue;
                    }
                }
            }
            break;
        }

        result
    }

    fn normalize_whitespace(&self, text: &str) -> String {
        let mut result = String::new();
        let mut last_was_space = false;
        for ch in text.chars() {
            if ch == ' ' || ch == '\t' {
                if !last_was_space {
                    result.push(' ');
                    last_was_space = true;
                }
            } else if ch == '\n' {
                result.push('\n');
                last_was_space = false;
            } else {
                result.push(ch);
                last_was_space = false;
            }
        }
        // Collapse multiple newlines
        while result.contains("\n\n\n") {
            result = result.replace("\n\n\n", "\n\n");
        }
        result
    }

    pub fn input_type(&self) -> TypeSpec {
        TypeSpec { type_name: "string".to_string(), nullable: true }
    }

    pub fn output_type(&self) -> TypeSpec {
        TypeSpec { type_name: "string".to_string(), nullable: true }
    }
}
